Lms/generate teacher notifications

anathomical opened the pull request:
## WHAT
Add code to actually generate `TeacherNotification` records so that they become available for email rollups.  Also add front-end access to the ability to change email rollup configuration so that teachers can control what they get and how often.

NOTE: I think there's a strong argument to be made to do the backend work and the front-end work in separate PRs, but launching one without the other doesn't make sense since we don't want to start sending emails without a way for teachers to control what they're getting, but we also don't want to imply that they can control message settings before they can actually get anything.  So if people want to review just the front-end or just the back-end of this PR, I think that could be helpful.
## WHY
We want teachers to be able to select specific student-related events to be notified of, and also to control how often they receive those notifications.  Once a teacher has their configuration picked (or if they leave the default settings in place), we also want them to actually receive roll-up emails of those notifications.
## HOW
- Add and update `Student` concern to have powerful utility methods: `assigned_activities` and `incomplete_assigned_activities` to help with later code to determine which notifications are relevant for a given student
- Hook up a Sidekiq worker to the `ActivitySession` completion code that determines which notifications a given activity completion might qualify for
- Have that worker write new `TeacherNotification` records for any relevant teachers
- Update the Teacher Account settings front-end to allow a teacher to change which student events they want to get notifications for, and how often they want to receive an email rollup of those notifications

### Screenshots
![image](https://user-images.githubusercontent.com/331565/236552927-87e52c1a-5267-4190-a1bb-72b289a67379.png)

### Notion Card Links
https://www.notion.so/quill/Send-a-roll-up-email-of-student-related-events-to-teachers-at-a-frequency-of-their-choosing-a074aabc41eb41be84f732db7a63aa49?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | Yes

anathomical committed:
WIP front-end notification setting interface

anathomical committed:
Add a hook to the ActivitySession controller to send notifications

When an activity session is complete queue a job to check to see what notifications might be relevant and send them to the appropriate teachers

anathomical committed:
Add some additional ActiveRecord relationships

These are there to facilitate some new ActiveRecord joins and includes

anathomical committed:
Update relationships in Student concern to find assigned activities

Update the logic in the assigned_activities relationship definition, and add a way to filter for incomplete_assigned_activities

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/generate-teacher-notifications

anathomical committed:
Add specs for the code that figures out which notifications to send

anathomical committed:
Fix linting errors

anathomical committed:
Make sure not to send Notifications of types teachers don't want

If a teacher does not have a TeacherNotificationSetting for a specific message type, don't send that type to them

anathomical committed:
Update the TeacherAccount editor to set email notification configs

anathomical committed:
Do some style tweaks to bring appearance in line with Figma specs

anathomical committed:
Fix lint

anathomical committed:
Fix specs

anathomical committed:
Refactor to reduce cyclomatic complexity to pass linters

anathomical committed:
Fix a linting indentation error

brendanshean commented on services/QuillLMS/app/controllers/api/v1/teacher_notification_settings_controller.rb:
```diff
@@ -32,9 +32,11 @@ def bulk_update
-    TeacherNotificationSetting.notification_types.to_h do |notification_type|
-      [notification_type, @current_settings.exists?(notification_type: notification_type)]
-    end
+    {
```
Since current_settings is a TeacherNotificationSettings collection, you could extract it into a class method within `TeacherNotificationSetting` and store the `exists?` as a temp to limit the number of database calls.

app/models/teacher_notification_setting
```
def self.blah
  temp = pluck(:notification_type).uniq

  notification_types.to_h do |notification_type|
    [notification_type, notification_type.in?(temp)]
  end
end
```

Then this becomes

```
private def types_and_values
  { teacher_notification_settings: @current_settings.blah }
end
```

brendanshean commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -638,15 +638,13 @@ def generate_teacher_account_info
+    user_attributes[:teacher_notification_settings] = teacher_notification_settings_info
     user_attributes[:subject_area_ids] = subject_area_ids
 
-    if school && school.name
```
Thanks for cleaning this up!

brendanshean commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -873,5 +871,23 @@ def school_or_district_premium?
   end
+
+  private def teacher_notification_settings_info
+    TeacherNotificationSetting.notification_types.to_h do |notification_type|
```
You can reuse the method discussed in the comment on teacher_notification_settings_controller.rb

brendanshean commented on services/QuillLMS/app/models/concerns/student.rb:
```diff
@@ -18,9 +18,28 @@ module Student
 
     has_many :activity_sessions, dependent: :destroy
-    has_many :assigned_activities, through: :classrooms, source: :activities
+    has_many :assigned_activities, -> { where("students_classrooms.student_id = ANY (classroom_units.assigned_student_ids)") }, through: :classrooms, source: :activities
```
Great job figuring out that the previous relation was incorrect.

brendanshean commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+
+    private def send_complete_diagnostic
+      return unless @activity_session.activity.classification.key == ActivityClassification::DIAGNOSTIC_KEY
```
What about: `return unless @activity_session.is_diagnostic?`

brendanshean commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,82 @@
+      return if unit_template.nil? || unit_template.recommendations.empty?
+
+      # Return early if the user has never completed a Diagnostic since any activities that they might be assigned can't be from recommendations in that case
+      return unless @activity_session.user.activity_sessions
```
What about moving lines 44-48 to user.rb

```
def completed_diagnostics?
...
end
```

This allows removal of the comment in line 44 and then just
`return unless @activity_session.user.completed_diagnostics?`

brendanshean commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,82 @@
+
+      # Return early if the user has any incomplete assigned activities that are recommendations
+      # Note that if the student has recommendations from two different diagnostics, they have to finish all recommendations from BOTH or this will returrn early
+      return if @activity_session.user.incomplete_assigned_activities
```
Similarly, what about adding a new scope in user.rb:

def incomplete_recommended_activities?
...
end

brendanshean replied:
Also, can the  .where..distinct.count > 0 could be rewritten as just `exists?`

brendanshean commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+
+    private def send_complete_all_assigned_activities
+      return if @activity_session.user.incomplete_assigned_activities
```
Instead of .where.count, what about exists?

brendanshean reviewed (approved):
I did a pass of the backend and it looked good to me.  Nice work on finishing up this last piece.

brendanshean reviewed (approved):
I did a pass of the backend and it looked good to me.  Nice work on finishing up this last piece.

eadams17 commented on services/QuillLMS/client/app/bundles/Teacher/components/accounts/edit/teacher_email_notifications.jsx:
```diff
@@ -1,36 +1,172 @@
-  render() {
+  function renderNewsletterCheckbox() {
+    const selectedClass = (sendNewsletter) ? "selected" : "unselected"
+    const checkboxImg = (sendNewsletter) ? (<img alt="check" src={smallWhiteCheckSrc} />) : ""
```
small nit, but I think we can do without the booleans or `img` element wrapped in parens

eadams17 commented on services/QuillLMS/client/app/bundles/Teacher/components/accounts/edit/teacher_email_notifications.jsx:
```diff
@@ -1,36 +1,172 @@
+  // in testing that during render cycles where "passed" values updated, the useState
+  // calls weren't updating, so the UI would revert to the original state until refresh.
+  // I suspect that it might be because the save call makes three separate API calls,
+  // but at least adding these hooks makes things work.
```
So I think what you're trying to achieve with using passed values for the initial state that will update on re-render without needing to use `useEffect` can be achieved with passing a key prop to this `TeacherEmailNotifications` component-- this post goes into a bit more detail of how that works: https://tkdodo.eu/blog/putting-props-to-use-state

anathomical committed:
Make minor tweaks to email templates based on product feedback

anathomical committed:
Make some tweaks based on code review

anathomical committed:
Make updates to the settings interface based on product feedback

anathomical committed:
Make some backend changes based on PR feedback

anathomical committed:
Do some more refactoring based on PR feedback

anathomical committed:
Make a small tweak to simplify ignoring empty snackbar values

dandrabik commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -873,5 +871,21 @@ def school_or_district_premium?
+  private def school_account_info
+    return school if school&.name
+
+    School.find_by_name(School::NO_SCHOOL_SELECTED_SCHOOL_NAME)
```
Let's use `School.find_by(name: School::NO_SCHOOL_SELECTED_SCHOOL_NAME)`. (preferring `Model.find_by(attribute:)` is in a TODO Rubocop rule we haven't turned on yet).

dandrabik commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,86 @@
+    private def send_complete_all_diagnostic_recommendations
+      unit_template = @activity_session.unit_template
+      # Return early if the activity that was just completed wasn't a recommendation
+      return if unit_template.nil? || unit_template.recommendations.empty?
```
Consider `return unless unit_template&.recommendations&.present?`

dandrabik commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,86 @@
+      end
+    end
+
+    private def teacher_receives_type(user, type)
```
This feels like a method that could be on user, e.g. `receives_notification_type?(type)`

dandrabik commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,86 @@
+      @activity_session = find_activity_session(activity_session_id)
+      return unless @activity_session
+
+      TeacherNotification.transaction do
```
Does this need to be wrapped in a transaction? We have a lot of DB contention on locks, so wondering if this could be removed. Maybe consider breaking into 3 jobs if the concern is keeping the job idempotent.

anathomical replied:
This is a good point.  Combined with your note below about expansion, I think it's probably a good idea to split it into multiple workers.  That said, each of the workers will be in a position where they write a variable number of records (based on how many teachers and co-teachers there are for the class an even takes place in), and I'm a little worried about one of those records saving, the job crashing, then re-running and doubling notifications for one of the teachers but not the other.  Or getting stuck in a retry loop where it keeps adding a record to one teacher, but the other always fails.  Maybe it's better to just take that risk than to add a transaction block, but that's what was on my mind when I added the transaction.

Thoughts?

dandrabik commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,86 @@
+        .first
+    end
+
+    private def send_complete_diagnostic
```
Note, every method in this file has `activity_session` as its base object. Take or leave, you could consider moving some methods there, e.g. `activity_session.send_complete_diagnostic_notification`

dandrabik commented on services/QuillLMS/app/workers/teacher_notifications/send_notification_worker.rb:
```diff
@@ -0,0 +1,86 @@
+      TeacherNotification.transaction do
+        send_complete_diagnostic
+        send_complete_all_diagnostic_recommendations
+        send_complete_all_assigned_activities
```
Another question. We'll be adding more of these notifications in the future. Is the plan to add them all in this file? That might make testing and maintenance more complicated than breaking them up into separate jobs.

dandrabik reviewed (approved):
Left some style notes and considerations, but nothing blocking approval. Thanks for pushing this along!

anathomical committed:
A couple of final tweaks based on product feedback

anathomical committed:
Remove <li> from the truncation partial

anathomical committed:
Move some logic into user model

anathomical committed:
Begin refactor to more of a fanout worker pattern for sending notifications

anathomical committed:
Refactor notification creation jobs to run in separate jobs

This makes it cleaner to test and easier to expand without the code becoming too complicated to get your head around

anathomical committed:
Fix lint errors

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/generate-teacher-notifications
